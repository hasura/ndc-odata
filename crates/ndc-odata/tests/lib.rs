use serde::Deserialize;
use std::path;

#[derive(Deserialize)]
struct Test {
    endpoint: String,
    method: Method,
    data: serde_json::Value,
}

#[derive(Deserialize)]
enum Method {
    GET,
    POST,
}

#[test_each::path(glob = "crates/ndc-odata/tests/**/*.json")]
fn test_snapshot(path: path::PathBuf) {
    let client = reqwest::blocking::Client::new();
    let content = std::fs::read_to_string(&path).unwrap();
    let test: Test = serde_json::from_str(&content).unwrap();

    let request = match test.method {
        Method::GET => client.get(test.endpoint),
        Method::POST => client.post(test.endpoint).json(&test.data),
    };

    // The path to the root of the `ndc-odata` crate.
    // Removing this means that we have a name that is consistent across platforms.
    let root = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));

    insta::with_settings!({
        input_file => &path,
        snapshot_path => root.join("tests/snapshots"),
    }, {
        insta::assert_json_snapshot!(
            path.strip_prefix(root).unwrap().to_str().unwrap(),
            request.send().unwrap().json::<serde_json::Value>().unwrap()
        )
    });
}
