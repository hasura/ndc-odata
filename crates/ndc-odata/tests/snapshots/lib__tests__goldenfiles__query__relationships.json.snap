---
source: crates/ndc-odata/tests/lib.rs
expression: "request.send().unwrap().json::<serde_json::Value>().unwrap()"
input_file: crates/ndc-odata/tests/goldenfiles/query/relationships.json
---
[
  {
    "rows": [
      {
        "first_name": "Russell",
        "last_name": "Whyte",
        "best_friend": {
          "first_name": "Scott",
          "last_name": "Ketchum"
        }
      },
      {
        "first_name": "Scott",
        "last_name": "Ketchum",
        "best_friend": {
          "first_name": "Russell",
          "last_name": "Whyte"
        }
      },
      {
        "first_name": "Ronald",
        "last_name": "Mundy",
        "best_friend": {
          "first_name": "Scott",
          "last_name": "Ketchum"
        }
      },
      {
        "first_name": "Javier",
        "last_name": "Alfred",
        "best_friend": {
          "first_name": "Vincent",
          "last_name": "Calabrese"
        }
      },
      {
        "first_name": "Willie",
        "last_name": "Ashmore",
        "best_friend": {
          "first_name": "Vincent",
          "last_name": "Calabrese"
        }
      },
      {
        "first_name": "Vincent",
        "last_name": "Calabrese",
        "best_friend": {
          "first_name": "Javier",
          "last_name": "Alfred"
        }
      },
      {
        "first_name": "Clyde",
        "last_name": "Guess",
        "best_friend": {
          "first_name": "Keith",
          "last_name": "Pinckney"
        }
      },
      {
        "first_name": "Keith",
        "last_name": "Pinckney",
        "best_friend": {
          "first_name": "Clyde",
          "last_name": "Guess"
        }
      },
      {
        "first_name": "Marshall",
        "last_name": "Garay",
        "best_friend": {
          "first_name": "Keith",
          "last_name": "Pinckney"
        }
      },
      {
        "first_name": "Ryan",
        "last_name": "Theriault",
        "best_friend": {
          "first_name": "Elaine",
          "last_name": "Stewart"
        }
      },
      {
        "first_name": "Elaine",
        "last_name": "Stewart",
        "best_friend": {
          "first_name": "Ryan",
          "last_name": "Theriault"
        }
      },
      {
        "first_name": "Sallie",
        "last_name": "Sampson",
        "best_friend": {
          "first_name": "Joni",
          "last_name": "Rosales"
        }
      },
      {
        "first_name": "Joni",
        "last_name": "Rosales",
        "best_friend": {
          "first_name": "Sallie",
          "last_name": "Sampson"
        }
      },
      {
        "first_name": "Georgina",
        "last_name": "Barlow",
        "best_friend": {
          "first_name": "Angel",
          "last_name": "Huffman"
        }
      },
      {
        "first_name": "Angel",
        "last_name": "Huffman",
        "best_friend": {
          "first_name": "Georgina",
          "last_name": "Barlow"
        }
      },
      {
        "first_name": "Laurel",
        "last_name": "Osborn",
        "best_friend": {
          "first_name": "Sandy",
          "last_name": "Osborn"
        }
      },
      {
        "first_name": "Sandy",
        "last_name": "Osborn",
        "best_friend": {
          "first_name": "Laurel",
          "last_name": "Osborn"
        }
      },
      {
        "first_name": "Ursula",
        "last_name": "Bright",
        "best_friend": {
          "first_name": "Krista",
          "last_name": "Kemp"
        }
      },
      {
        "first_name": "Genevieve",
        "last_name": "Reeves",
        "best_friend": {
          "first_name": "Ursula",
          "last_name": "Bright"
        }
      },
      {
        "first_name": "Krista",
        "last_name": "Kemp",
        "best_friend": {
          "first_name": "Ursula",
          "last_name": "Bright"
        }
      }
    ]
  }
]
